src/remote.c

```diff
@@ -670,21 +670,46 @@ int git_remote_set_pushurl(git_repository *repo, const char *remote, const char*
 	return set_url(repo, remote, CONFIG_PUSHURL_FMT, url);
 }
 
-const char* git_remote__urlfordirection(git_remote *remote, int direction)
+static int resolve_url(git_buf *resolved_url, const char *url, int direction, const git_remote_callbacks *callbacks)
 {
-	assert(remote);
+	int status;
 
-	assert(direction == GIT_DIRECTION_FETCH || direction == GIT_DIRECTION_PUSH);
+	if (callbacks && callbacks->resolve_url) {
+		git_buf_clear(resolved_url);
+		status = callbacks->resolve_url(resolved_url, url, direction, callbacks->payload);
+		if (status != GIT_PASSTHROUGH) {
+			git_buf_sanitize(resolved_url);
```

Comment:
Is this needed? Implementations of this callback only have `git_buf_set` available, which will always NUL-terminate the buffer. If the callback implementation does nothing at all, then `git_buf_clear` will already have the buffer NUL-terminated.

Reply:
My initial round at this was accessing the ptr directly. I don't have strong opinions, but I think it makes sense as a "boundary" measure: any git_buf we allow userspace to see must be sanitized, just in case (my initial implementation forgot to increment the buffer size, for example).

Reply:
Oh, right. I thought that `git_buf` was completely opaque to users of libgit2, but in fact it isn't. So it's fine as-is, I guess

Reply:
Heh, I had the exact inverse feeling 😮. A public struct with no function at all. Turns out I was wrong, and I've been listening to a synthetized choir of tibetan monks reading the header contents out loud for 2 days now. It's quite melodic 😉.

Reply:
It's definitely true that `git_buf` is somewhat strangely exposed. What I find most interesting is that we do not even provide any way to correctly initialize a `git_buf` except for saying `git_buf buf = {0}`.

Reply:
Yeah, I'd say because exposing the `__const` OOM buf would be weird ? Anyways, I think this is okay as is, maybe something to clarify in the develop guide…